consensus/clique/clique_test.go

```diff
@@ -113,3 +118,395 @@ func TestReimportMirroredState(t *testing.T) {
+			// SCENARIO-1
+			// signers: A..H (8 count)
+			//
+			/*
+				Step 1
+				A fully in-order chain exists and validator 8 has just produced an in-turn block.
+
+				1A x
+				2B  x
+				3C   x
+				4D    x
+				5E     x
+				6F      x
+				7G       x
+				8H        x
+
+				Step 2
+				... and then validators 5, 7, and 8 go offline.
+
+				1A x
+				2B  x
+				3C   x
+				4D    x
+				5E     x   -
+				6F      x
+				7G       x -
+				8H        x-
+
+				Step 3
+				Two forks form, one with an in-order block from validator 1
+				and then an out of order block from validator 3.
+
+				The second fork forms from validators 2, 4, and 6 in order.
+				Both have a net total difficulty of 3 more than the common ancestor.
+
+				1A x        y
+				2B  x       z
+				3C   x       y
+				4D    x      z
+				5E     x   -
+				6F      x     z
+				7G       x -
+				8H        x-
+			*/
+			lenSigners:   8,
+			commonBlocks: []int{1, 2, 3, 4, 5, 6, 7, 0, 1, 2, 3, 4, 5, 6, 7},
+			forks: [][]int{
+				{1, 3, 5}, // 2, 4, 6
+				{0, 2},    // 1, 3
+			},
+			canonicalForkIndex: func(forkHeads ...*types.Header) int {
+				// Prefer the shorter fork.
+				minHeight := math.MaxBig63.Uint64()
+				n := -1
+				for i, head := range forkHeads {
+					if h := head.Number.Uint64(); h < minHeight {
+						n = i
+						minHeight = h
+					}
+				}
+				return n
+			},
+			assertions: []func(chain *core.BlockChain, forkHeads ...*types.Header){
+				// Assert that the net total difficulties of each fork are equal.
+				func(chain *core.BlockChain, forkHeads ...*types.Header) {
+					d := new(big.Int)
+					for i, head := range forkHeads {
+						td := chain.GetTd(head.Hash(), head.Number.Uint64())
+						if i == 0 {
+							d.Set(td)
+							continue
+						}
+						if d.Cmp(td) != 0 {
+							t.Fatalf("want equal fork heads total difficulty")
+						}
+					}
+				},
+			},
+		},
+		{
+			// SCENARIO-2
+
+			/*
+					Step 1
+					For the second scenario with the same validator set and in-order chain with
+					validator 7 having just produced an in order block, then validators 7 and 8 go offline.
+
+					1A x
+					2B  x
+					3C   x
+					4D    x
+					5E     x
+					6F      x
+					7G       x
+					8H
+
+					1A x
+					2B  x
+					3C   x
+					4D    x
+					5E     x
+					6F      x
+					7G       x-
+					8H        -
+
+					Two forks form, 1,3,5 on one side and 2,4,6 on the other.
+					Both forks become aware of the other fork after producing their third block.
+					In this case both forks have equal total difficulty and equal length.
+
+					1A x       x
+					2B  x      y
+					3C   x      x
+					4D    x     y
+					5E     x     x
+					6F      x    y
+					7G       x-
+					8H        -
+
+				FIXME(meowsbits): This scenario yields a "recently signed" error
+				when attempting to import Signer 5 (really #6 b/c zero-indexing) into the
+				second fork.
+
+				On that fork, the sequence of signers is specified to be
+				... 7, 0, 1, 2, 3, 4, 5, 6, 1, 3, 5
```

Comment:
I'm reading this as signer `5` should not actually be allowed to validate this block, since it's eg. `17-13=4` blocks from their last validation, where the limit should be `8/2+1=5`. This would mean that the `error recently signed` coming from the test for this scenario would be right.